### PR TITLE
[serial]: Added a new method to SerialEvents trait

### PR DESCRIPTION
**Pull request by AlexandruCihodaru**
- Added in_buffer_ready methog to SerialEvents triat. This method
has the role to inform the vmm that there are no pending bytes in
the in_buffer container and it shoudl do something about it(i.e
provision it with more bytes from the input source).

Signed-off-by: AlexandruCihodaru <cihodar@amazon.com>
Suggested-by: Andreea Florescu <fandree@amazon.com>

### CODE REVIEW

File: src/serial.rs [hunk 1]
```diff
@@ -117,6 +117,8 @@ pub trait SerialEvents {
     fn out_byte(&self);
     /// An error occurred while writing a byte to serial output resulting in a lost byte.
     fn tx_lost_byte(&self);
+    /// Let vmm know that in_buffer is empty.
```
Comment (lauralt) [inline src/serial.rs:120]: Can we rephrase this to simply: "The input buffer is empty."?

File: src/serial.rs [hunk 2]
```diff
@@ -450,12 +457,13 @@ impl<T: Trigger, EV: SerialEvents, W: Write> Serial<T, EV, W> {
                 // interrupt identification register and RDA bit when no
                 // more data is available).
                 self.del_interrupt(IIR_RDA_BIT);
-                if self.in_buffer.len() <= 1 {
+                let byte = self.in_buffer.pop_front().unwrap_or_default();
+                if self.in_buffer.is_empty() {
                     self.clear_lsr_rda_bit();
+                    self.events().in_buffer_empty();
```
Comment (lauralt) [inline src/serial.rs:463]: Any reason to use `events()` instead of `events` field?

File: src/serial.rs [hunk 3]
```diff
@@ -586,6 +594,31 @@ mod tests {
...
+        }
+    }
+
+    impl Default for ExampleSerialMetrics {
+        fn default() -> Self {
+            ExampleSerialMetrics {
+                read_count: AtomicU64::new(0),
+                out_byte_count: AtomicU64::new(0),
+                tx_lost_byte_count: AtomicU64::new(0),
+                buffer_ready_event: Some(EventFd::new(libc::EFD_NONBLOCK).unwrap()),
```
Comment (lauralt) [inline src/serial.rs:619]: I would expect the `Default` for an Option<something> to be None. Can we initialize it with `Some()` only in the test that is checking the `in_buffer_empty()`? This way you can keep the `#[derive(Default)]` above as well.

File: src/serial.rs [hunk 4]
```diff
@@ -566,11 +574,11 @@ mod tests {
         }
     }
 
-    #[derive(Default)]
     struct ExampleSerialMetrics {
```
Comment (lauralt) [inline src/serial.rs:577]: `Metrics` is no longer accurate. Can we rename it to `Events`?

File: src/serial.rs [hunk 5]
```diff
@@ -840,6 +873,22 @@ mod tests {
         assert_eq!(serial.events.read_count.count(), 1);
         assert_eq!(serial.events.out_byte_count.count(), 1);
         assert_eq!(serial.events.tx_lost_byte_count.count(), 1);
+
+        // This DATA read should cause the `SerialEvents::buffer_read` method to be invoked.
+        // And since it was the last byte from in buffer the `SerialEvents::in_buffer_empty`
```
Comment (lauralt) [inline src/serial.rs:878]: Can we also test that `in_buffer_empty` is not called if there are more bytes in the buffer?

**Review by lauralt — COMMENTED** [review]
There are a couple of typos in the commit message (also `in_buffer_ready` -> `in_buffer_empty`). Can you please fix them as well?

File: src/serial.rs [hunk 6]
```diff
@@ -117,6 +117,8 @@ pub trait SerialEvents {
     fn out_byte(&self);
     /// An error occurred while writing a byte to serial output resulting in a lost byte.
     fn tx_lost_byte(&self);
+    /// The input buffer is empty.
```
Comment (andreeaflorescu) [inline src/serial.rs:120]: I'd like to see more in the doc comment here. 

I think it's useful to let people know how this can be used. For example, we can write something in the lines of: "This event can be used by the consumer to re-enable events coming from the serial input."

File: src/serial.rs [hunk 7]
```diff
@@ -586,6 +606,20 @@ mod tests {
...
+            match self
+                .buffer_ready_event
+                .as_ref()
+                .map_or(Ok(()), |buf_ready| buf_ready.write(1))
+            {
+                Ok(_) => (),
+                Err(err) => println!(
+                    "Could not signal that serial device buffer is ready: {:?}",
+                    err
+                ),
```
Comment (andreeaflorescu) [inline src/serial.rs:620]: In tests we should use unwraps because otherwise we might end up with silent fails.

File: src/serial.rs [hunk 8]
```diff
@@ -812,9 +846,18 @@ mod tests {
...
     fn test_serial_events() {
         let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
-        let metrics = Arc::new(ExampleSerialMetrics::default());
+
+        let events_ = Arc::new(ExampleSerialEvents::new());
         let mut oneslot_buf = [0u8; 1];
-        let mut serial = Serial::with_events(intr_evt, metrics, oneslot_buf.as_mut());
+        let mut serial = Serial::with_events(intr_evt, events_, oneslot_buf.as_mut());
+
+        assert!(serial
```
Comment (andreeaflorescu) [inline src/serial.rs:854]: Can you add a comment to explain why we expect this to be an error?

File: src/serial.rs [hunk 9]
```diff
@@ -567,13 +575,25 @@ mod tests {
...
+        buffer_ready_event: Option<EventFd>,
+    }
+
+    impl ExampleSerialEvents {
+        fn new() -> Self {
+            ExampleSerialEvents {
+                read_count: AtomicU64::new(0),
+                out_byte_count: AtomicU64::new(0),
+                tx_lost_byte_count: AtomicU64::new(0),
+                buffer_ready_event: Some(EventFd::new(libc::EFD_NONBLOCK).unwrap()),
```
Comment (andreeaflorescu) [inline src/serial.rs:591]: I don't see the need to have `buffer_ready_event` as an option. Why can't we just always initialize it with `EventFd::new(libc::EFD_NONBLOCK).unwrap()`? This would simplify the code where `buffer_ready_event` is used, and where that field is not used, then having it initialize doesn't cause any problems.

File: src/serial.rs [hunk 10]
```diff
@@ -812,9 +846,18 @@ mod tests {
...
-        let mut serial = Serial::with_events(intr_evt, metrics, oneslot_buf.as_mut());
+        let mut serial = Serial::with_events(intr_evt, events_, oneslot_buf.as_mut());
+
+        assert!(serial
+            .events
+            .buffer_ready_event
+            .as_ref()
+            .unwrap()
+            .read()
+            .is_err());
```
Comment (andreeaflorescu) [inline src/serial.rs:860]: Let's check for the actual error instead of just that a generic error is returned. In this case, because io::Error does not implement `PartialEq`, we can use `.kind` as follows:
```
        assert_eq!(
            serial
                .events
                .buffer_ready_event
                .as_ref()
                .unwrap()
                .read()
                .unwrap_err()
                .kind(),
            ErrorKind::WouldBlock
        );
```

File: src/serial.rs [hunk 11]
```diff
@@ -840,6 +893,31 @@ mod tests {
...
+        );
+        let _res = serial.enqueue_raw_bytes(&[1, 2]);
+        serial.read(DATA_OFFSET);
+        assert!(serial
+            .events
+            .buffer_ready_event
+            .as_ref()
+            .unwrap()
+            .read()
+            .is_err());
```
Comment (andreeaflorescu) [inline src/serial.rs:920]: Compare against the actual error instead of using `is_err`.

**Review by lauralt — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]
